File: #1239.maximum-length-of-a-concatenated-string-with-unique-characters.cpp
class Solution {
public:
    int maxLength(vector<string>& arr) {

        size = arr.size();
        string s;
        maxLength(arr, 0, s);
        return max;
    }

private:
    int size = 0;
    int max = 0;

    void maxLength(vector<string> arr,
                   int start,
                   string &s) {

        if (s.length() > max) {
            max = s.length();
        }

        if (start == size) {
            return;
        }

        for (int i = start; i < size; i++) {
            string temp = s + arr[i];
            bool mask[27] = {false};
            for (char c: s) {
                mask[c-'a'] = true;
            }

            bool unique_chars = true;
            for (char c: arr[i]) {
                if (mask[c-'a']) {
                    unique_chars = false;
                    break;
                }
                mask[c-'a'] = true;
            }

            if (unique_chars) {
                maxLength(arr, i + 1, temp);
            }
        }
    }
};

File: #1239.maximum-length-of-a-concatenated-string-with-unique-characters.js
/**
 * @param {string[]} arr
 * @return {number}
 */
var maxLength = function(arr) {
    const isValid = (str) => str.length === new Set(str).size;
    const valid = arr.filter(s => isValid(s));
    let combinations = valid;
    for (let i = 0; i < valid.length; i++) {
        combinations = combinations.flatMap(c => {
            if (!isValid(c + valid[i])
                || c + valid[i] > valid[i] + c) {
                return c;
            } else {
                return [c, c + valid[i]];
            }
        });
    }
    return Math.max(...combinations.map((c => c.length)), 0);
};
